node/pkg/cni/token_watch.go

```diff
@@ -1,56 +1,158 @@
+}
+
+type TokenUpdate struct {
+	Token          string
+	ExpirationTime time.Time
+	Error          error
+}
+
+func NewTokenRefresher(namespace string, serviceAccountName string) *TokenRefresher {
+	return &TokenRefresher{
+		tokenSupported:     false,
+		tokenOnce:          &sync.Once{},
+		namespace:          namespace,
+		serviceAccountName: serviceAccountName,
+		tokenChan:          make(chan TokenUpdate),
+	}
+}
+
+func (t *TokenRefresher) UpdateToken() TokenUpdate {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		logrus.WithError(err).Error("Error generating kube config.")
+		return tokenUpdateError(err)
+	}
+	err = rest.LoadTLSFiles(cfg)
+	if err != nil {
+		logrus.WithError(err).Error("Error loading TLS files.")
+		return tokenUpdateError(err)
+	}
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to create clientset for CNI kubeconfig")
+		return tokenUpdateError(err)
+	}
+
+	validity := int64(defaultCNITokenValiditySeconds)
+	tr := &authenticationv1.TokenRequest{
+		Spec: authenticationv1.TokenRequestSpec{
+			Audiences:         []string{"kubernetes"},
+			ExpirationSeconds: &validity,
+		},
+	}
+
+	tokenRequest, err := clientset.CoreV1().ServiceAccounts(t.namespace).CreateToken(context.TODO(), t.serviceAccountName, tr, metav1.CreateOptions{})
+	if apierrors.IsNotFound(err) && !t.tokenRequestSupported(clientset) {
+		logrus.WithError(err).Error("Unable to create token for CNI kubeconfig as token request api is not supported")
+		return tokenUpdateError(errors.New("token request api not supported"))
+	}
 	if err != nil {
-		panic(err)
+		logrus.WithError(err).Error("Unable to create token for CNI kubeconfig")
+		return tokenUpdateError(err)
 	}
-	defer watcher.Close()
 
-	// Watch for changes to the serviceaccount directory. Rety if necessary.
+	return TokenUpdate{
+		Token:          tokenRequest.Status.Token,
+		ExpirationTime: tokenRequest.Status.ExpirationTimestamp.Time,
+	}
+}
+
+func (t *TokenRefresher) TokenChan() <-chan TokenUpdate {
+	return t.tokenChan
+}
+
+func (t *TokenRefresher) Run() {
+	var nextExpiration time.Time
 	for {
-		if err := watcher.Add(serviceaccountDirectory); err != nil {
-			// Error watching the file - retry
-			logrus.WithError(err).Error("Failed to watch Kubernetes serviceaccount files.")
-			time.Sleep(5 * time.Second)
-			continue
+		tu := t.UpdateToken()
+		if tu.Error == nil {
+			nextExpiration = tu.ExpirationTime
+			t.tokenChan <- tu
+		}
+		now := time.Now()
+		var sleepTime time.Duration
+		// Do some basic rate limiting to prevent flooding the kube apiserver with requests
+		if nextExpiration.Before(now.Add(minTokenRetryDuration * defaultRefreshFraction)) {
+			sleepTime = minTokenRetryDuration
+		} else {
+			sleepTime = nextExpiration.Sub(now) / defaultRefreshFraction
 		}
+		if logrus.IsLevelEnabled(logrus.DebugLevel) {
+			logrus.Debugf("Going to sleep for %s", sleepTime.String())
+		}
+		time.Sleep(sleepTime)
```

Comment:
Ok, so I think the "solution" to the stampede problem I was concerned with in the other comment is likely to just add some jitter to the sleep here so that we minimize the chance that all the calico/node pods try to hit the token request API at the same time.

I think it should be safe to refresh the token at a random time between 1/4 and 1/2 of the validity period, for example.